frontend/packages/console-plugin-sdk/src/withExtensions.tsx

```diff
@@ -45,50 +41,57 @@ import { Extension, ExtensionTypeGuard } from './typings';
  *
  * @param typeGuardMapper Object that maps prop names to extension type guards.
  * It's basically an object-based analogy to Redux `mapStateToProps` function.
+ *
+ * @returns `withExtensions` higher-order component creator function.
  */
 export const withExtensions = <
   TExtensionProps extends ExtensionProps<E>,
-  E extends Extension = Extension
+  E extends Extension = Extension,
+  TCombinedProps extends TExtensionProps = TExtensionProps
 >(
   typeGuardMapper: ExtensionTypeGuardMapper<E, TExtensionProps>,
-): (<P extends TExtensionProps>(
-  C: React.ComponentType<P>,
-) => React.ComponentType<Omit<P, keyof TExtensionProps>> & {
-  WrappedComponent: React.ComponentType<P>;
-}) => {
-  const allExtensions = pluginStore.getAllExtensions();
+): ((C: React.ComponentType<TCombinedProps>) => ExtensionHOC<TCombinedProps, TExtensionProps>) => (
+  WrappedComponent,
+) => {
+  const typeGuards = Object.values(typeGuardMapper);
+
+  if (typeGuards.length === 0) {
+    throw new Error('The object passed to withExtensions must contain at least one type guard');
+  }
+
+  const HOC: ExtensionHOC<TCombinedProps, TExtensionProps> = (props) => {
+    const extensionsInUse = useExtensions(...typeGuards);
 
-  // 1) Narrow extensions according to type guards
-  const matchedExtensions = _.flatMap(
-    Object.values(typeGuardMapper).map((tg) => allExtensions.filter(tg)),
-  );
+    const extensionProps = React.useMemo(
+      () =>
+        Object.keys(typeGuardMapper).reduce((acc, propName) => {
+          acc[propName] = extensionsInUse.filter(typeGuardMapper[propName]);
+          return acc;
+        }, {} as ExtensionProps<E>),
+      [extensionsInUse],
+    );
 
-  // 2.a) Compute flags relevant for gating matched extensions
-  const gatingFlagNames = getGatingFlagNames(matchedExtensions);
+    const combinedProps = React.useMemo(
```

Comment:
is there actually any benefit using `useMemo` ? Why not just `<WrappedComponent {...props} {...extensionProps} />`

Reply:
> Why not just `<WrappedComponent {...props} {...extensionProps} />`

`WrappedComponent` is `React.ComponentType<TCombinedProps>` so the props object needs to be of type `TCombinedProps`, otherwise TS compiler will complain.

@rawagner Agreed that there's no real benefit to `React.useMemo` here, but we still need to retain the `as TCombinedProps` cast and therefore create a separate (combined) props object:

```ts
const combinedProps = {
  ...props,
  ...extensionProps,
} as TCombinedProps;
```